### PR TITLE
Avoid race condition in ILMHistorySotre

### LINKED ISSUE #50353: [CI] TimeSeriesLifecycleActionsIT testHistoryIsWrittenWithFailure
A number of failures on the master branch in the last day:

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-windows-compatibility/os=windows-2012-r2/331/console
https://gradle-enterprise.elastic.co/s/clxgk64w274hw

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=ubuntu-18.04&&immutable/460/console
https://gradle-enterprise.elastic.co/s/ptsngwrgxzn64

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=centos-6&&immutable/460/console
https://gradle-enterprise.elastic.co/s/jdmkulxql5ris

```
java.lang.AssertionError: 

Expected: a value equal to or greater than <1>
     but: <0> was less than <1>

at __randomizedtesting.SeedInfo.seed([9BF20B84CA3FCF0E:49B89F3F80A0B0A0]:0)
at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
at org.junit.Assert.assertThat(Assert.java:956)
at org.junit.Assert.assertThat(Assert.java:923)
at org.elasticsearch.xpack.ilm.TimeSeriesLifecycleActionsIT.assertHistoryIsPresent(TimeSeriesLifecycleActionsIT.java:1157)
at org.elasticsearch.xpack.ilm.TimeSeriesLifecycleActionsIT.lambda$testHistoryIsWrittenWithDeletion$72(TimeSeriesLifecycleActionsIT.java:1103)
at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:866)
at org.elasticsearch.xpack.ilm.TimeSeriesLifecycleActionsIT.testHistoryIsWrittenWithDeletion(TimeSeriesLifecycleActionsIT.java:1102)
```

Does not reproduce:
```
gradlew ':x-pack:plugin:ilm:qa:multi-node:integTestRunner' --tests "org.elasticsearch.xpack.ilm.TimeSeriesLifecycleActionsIT.testHistoryIsWrittenWithSuccess" \
  -Dtests.seed=DA4311B28F41839E \
  -Dtests.security.manager=true \
  -Dtests.locale=id-ID \
  -Dtests.timezone=SystemV/MST7MDT \
  -Dcompiler.java=13
```

### PR DESCRIPTION
**Pull request by probakowski**
This change modifies ILMHistoryStore to always apply correct settings and mappings,
even if template is deleted and not yet recreated. This ensures that ILM history index
is correctly managed by ILM and also fixes flaky history tests that were prone to
triggering this race.

This commit also refactors and simplifies ILM history tests.

Closes #50353 and #52853

### CODE REVIEW

**Comment by probakowski** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by probakowski** [comment]
@elasticmachine update branch

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/history/ILMHistoryStore.java [hunk 1]
```diff
@@ -183,11 +188,19 @@ static void ensureHistoryIndex(Client client, ClusterState state, ActionListener
         if (ilmHistory == null && initialHistoryIndex == null) {
             // No alias or index exists with the expected names, so create the index with appropriate alias
             logger.debug("creating ILM history index [{}]", initialHistoryIndexName);
+
+            // Template below should be already defined as real index template but it can be deleted. To avoid race condition with its
+            // recreation we apply settings and mappings ourselves
+            byte[] templateBytes = TEMPLATE_ILM_HISTORY.loadBytes();
+            Map<String, Object> map = XContentHelper.convertToMap(new BytesArray(templateBytes, 0, templateBytes.length),
```
Comment (andreidan) [inline x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/history/ILMHistoryStore.java]: super minor nit: "templateAsMap" or a name along those lines would be a bit more readable perhaps?

**Review by andreidan — APPROVED** [review]
Thanks for fixing this @probakowski! LGTM

**Review by dakrone — APPROVED** [review]
This change looks good to be overall, but I think these tests will still end up failing due to the race condition between index creation via _bulk requests and manually being checked in the history store.

I think in order to fully fix them we need to add the settings to make the history store as non-async as possible when indexing, so we can prevent the alias/concrete-index issues that were causing failures in the past.

We can definitely merge this (thanks for doing this!), but I think we may still end up with some failures down the road.

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/history/ILMHistoryStore.java (hunk 1 above)
Comment (probakowski) [reply x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/history/ILMHistoryStore.java]: Good idea, changed

**Comment by probakowski** [comment]
Thanks @andreidan and @dakrone, I'll merge it as soon as build passes.
I think right now window for race is pretty small. Only possibility I see is that something deletes both history index and template between check in `ILMHistoryStore#ensureHistoryIndex` and actual indexing AND template is not recreated in time. Although not impossible it would be pretty hard to hit.
For what it's worth I've run 3000 back-to-back tests without single failure.

**Comment by probakowski** [comment]
@elasticmachine run elasticsearch-ci/bwc and run elasticsearch-ci/default-distro

**Comment by probakowski** [comment]
@elasticmachine update branch

**Pull request merged**